src/main.rs

```diff
@@ -1,4 +1,143 @@
-fn main() {
-    println!("Hello, world!");
+use axum::{
+    async_trait,
+    extract::{FromRequest, Path, RequestParts},
+    handler::Handler,
+    http::StatusCode,
+    response::{IntoResponse, Json, Response},
+    routing::get,
+    Router, Server,
+};
+use serde_json::Value;
+use std::collections::HashMap;
+
+#[derive(Debug)]
+enum Version {
+    V1,
+}
+
+#[derive(Debug)]
+struct JsonProxyError(reqwest::Error);
+
+#[tokio::main]
+async fn main() {
+    let meta_routes = Router::new()
+        .route("/", get(root_handler))
+        .route("/ping", get(heartbeat_handler));
+
+    let transit_routes = Router::new()
+        .route("/bus", get(transit_bus_handler))
+        .route("/modes", get(transit_modes_handler));
+
+    let api_routes = Router::new()
+        .route("/contacts", get(contacts_handler))
+        .route("/dictionary", get(dictionary_handler))
+        .route("/faqs", get(faqs_handler))
+        .route("/food/named/menu/pause", get(pause_menu_handler))
+        .route("/printing/color-printers", get(color_printers_handler))
+        .route("/spaces/hours", get(hours_handler))
+        .route("/tools/help", get(help_handler))
+        .nest("/transit", transit_routes)
+        .route("/webcams", get(webcams_handler));
+
+    let app = Router::new()
+        .nest("/", meta_routes)
+        .nest("/api/:version", api_routes)
+        .fallback(fallback.into_service());
+
+    Server::bind(&"0.0.0.0:3000".parse().unwrap())
+        .serve(app.into_make_service())
+        .await
+        .expect("server failed to exit successfully");
+}
+
+#[allow(clippy::unused_async)]
+async fn root_handler() -> Result<&'static str, StatusCode> {
+    Ok("Hello world!")
+}
+
+#[allow(clippy::unused_async)]
+async fn heartbeat_handler() -> Result<&'static str, StatusCode> {
+    Ok("pong")
+}
+
+#[allow(clippy::unused_async)]
+async fn fallback() -> impl IntoResponse {
+    (StatusCode::NOT_FOUND, "Not Found".to_string())
+}
+
+async fn gh_pages_handler(filename: &str) -> Result<Json<Value>, JsonProxyError> {
+    let url = format!("https://stodevx.github.io/AAO-React-Native/{}", filename).to_string();
+    let resp = request_handler(&url).await?;
+    Ok(resp)
 }
 
+macro_rules! gh_pages_handler {
+    ($name:ident,$filename:literal) => {
+        async fn $name(_version: Version) -> Result<Json<Value>, JsonProxyError> {
+            let data = gh_pages_handler($filename).await.unwrap();
+            Ok(data)
+        }
+    };
+}
+
+macro_rules! gh_pages_handlers {
+    ($([$name:ident, $filename:literal]),+ $(,)?) => {
+        $(
+            gh_pages_handler!($name, $filename);
+        )+
+    };
+}
+
+gh_pages_handlers!(
+    [contacts_handler, "contact-info.json"],
```

Comment:
If we wanted to represent the return type we get back from GitHub in terms of structs, we have the following schema from `data/_schemas` in AAO:

```yaml

type: object
additionalProperties: false
required: [title, buttonText, text, synopsis, category]
oneOf:
  - required: [phoneNumber]
  - required: [buttonLink]
properties:
  title: {type: string}
  phoneNumber: {$ref: '_defs.json#/definitions/phone'}
  buttonText: {type: string}
  buttonLink: {type: string}
  category: {type: string}
  synopsis: {type: string}
  image: {type: string}
  text: {type: string}
```

Note that some types (those built from directories) have a `data` top level key that maps to an array of subtypes (built [here](https://github.com/StoDevX/AAO-React-Native/blob/fdf012c8aafb77d646b55eb719bc63848dd8e761/scripts/bundle-data-dir.mjs#L40)).

Roughly translating this structure (and using `String` as the inner type for now), I get these two structs:

```rust
#[derive(serde::Serialize, serde::Deserialize)]
struct ContactInfoItem {
    title: String,
    #[serde(rename = "phoneNumber")]
    phone_number: Option<String>,
    #[serde(rename = "buttonText")]
    button_text: Option<String>,
    category: String,
    image: Option<String>,
    synopsis: String,
    text: String,
}

#[derive(serde::Serialize, serde::Deserialize)]
struct ContactInfoResponse {
    data: Vec<ContactInfoItem>,
}
```

These are possible types to pass as type parameters to `gh_pages_handler`! To verify that, you could pull this row out of the `gh_pages_handlers!` invocation temporarily and rewrite the generation using the `gh_pages_handler!` macro:

```rust
gh_pages_handler!(contacts_handler, "contact-info.json", ContactInfoResponse);
```

If we want to support using _either_ the blasé `Value` type or the custom `$response_type` version of the macro, we can do something like what we did with the trailing comma to the `gh_pages_handlers` macro, adding a `$(, $response_type:ty)?` to the `[ .. ]` expression and a corresponding `$(, $response_type)?` to the inner `gh_pages_handler!` invocation.

```rust
macro_rules! gh_pages_handlers {
    ($([$name:ident, $filename:literal $(, $response_type:ty)?]),+ $(,)?) => {
        $(
            gh_pages_handler!($name, $filename $(, $response_type)?);
        )+
    };
}
```

This way, if a row _doesn't_ have a `$response_type` (like the `dictionary_handler` row), it'll only pass the `$name` and `$filename` arguments.  If you do supply one, à la

```rust
    [contacts_handler, "contact-info.json", ContactInfoResponse],
```

then you will instead get a `gh_pages_handler` that returns a concrete `Result<Json<ContactInfoResponse>, JsonProxyError>` type.

Reply:
Note that in order to `derive` `serde::Serialize` and `serde::Deserialize`, you'll have to update the `serde` entry in the `dependencies` section of `Cargo.toml` to include the `derive` feature, which is not enabled by default:

```diff
-serde = "1.0.136"
+serde = { version = "1.0.136", features = ["derive"] }
```

Reply:
Cool, all of that looks to be working! I pulled out all of the untyped gh pages routes for now rather than maintaining two types of handlers so the rest will be added back now in the new format you suggested.